### PR TITLE
Feature/wcwpp 73 create queue worker

### PR DESCRIPTION
**Pull request by lucapircher**
## TLDR;

- adds the queue worker for retrying requests to GB Weiss
- refactors the order controller since a OrderRepository had to be introduced for the queue worker
- adds a helper class for interacting with wordpress api to allow easier mocking
- extracts the logic for sending a request to gb weiss into a command
- adds a last attemped date to the failed request to ensure that retries do not happen more frequently than every five minutes

## Additional Info

There is one incomplete test regarding authentication in the queue worker. This will be dealt with in a separate PR to not further increase the scope of this PR.

### CODE REVIEW

File: tests/Integration/FailedRequestRepositoryTest.php [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 namespace Tests\Integration;
 
+use DateTime;
 use Towa\GebruederWeissWooCommerce\Plugin;
 use Towa\GebruederWeissWooCommerce\FailedRequestQueue\FailedRequest;
 use Towa\GebruederWeissWooCommerce\FailedRequestQueue\FailedRequestRepository;
```
Comment (philipp-hemmer) [inline tests/Integration/FailedRequestRepositoryTest.php:8]: regarding line 24 in this file: i think this row should somehow get deleted after the test

File: tests/Integration/WordpressTest.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Tests\Integration;
+
+use Towa\GebruederWeissWooCommerce\Support\WordPress;
+
+class WordPressTest extends \WP_UnitTestCase
+{
+    public function test_it_can_sen_error_notifications_to_the_admin()
```
Comment (philipp-hemmer) [inline tests/Integration/WordpressTest.php]: typie in the title "send"

File: tests/Unit/RetryFailedRequestQueueWorkerTest.php [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        $worker = new RetryFailedRequestsQueueWorker($failedRequestRepository, $this->logisticsOrderFactory, $writeApi, $orderRepository, $this->settingsRepository);
+        $worker->start();
+
+        // wc-failed is the value returned by the settings repository
+        $order->shouldHaveReceived("set_status", ["wc-failed"]);
+    }
+
+    public function test_it_ensures_that_the_requests_are_authenticated()
+    {
+        $this->markTestIncomplete();
```
Comment (philipp-hemmer) [inline tests/Unit/RetryFailedRequestQueueWorkerTest.php:218]: there is no test here as far as i understand

File: tests/Integration/FailedRequestRepositoryTest.php (hunk 1 above)
Comment (lucapircher) [reply tests/Integration/FailedRequestRepositoryTest.php:8]: The wordpress testing framework cleans the database after each individual test. So all changes made during a test will be undone.

File: tests/Unit/RetryFailedRequestQueueWorkerTest.php (hunk 3 above)
Comment (lucapircher) [reply tests/Unit/RetryFailedRequestQueueWorkerTest.php:218]: I have updated the additional info section. I wanted to say that there is one incomplete test regarding authentication, and that will be addressed in a separate PR.

**Review by philipp-hemmer — APPROVED** [review]

File: includes/FailedRequestQueue/RetryFailedRequestsQueueWorker.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        while (true) {
+            $failedRequest = $this->failedRequestRepository->findOneToRetry();
+
+            if (is_null($failedRequest)) {
+                break;
+            }
+
+            $order = $this->orderRepository->findById($failedRequest->getOrderId());
+
+            if (is_null($order)) {
```
Comment (mattcrn) [inline includes/FailedRequestQueue/RetryFailedRequestsQueueWorker.php:99]: When would the order be null?

Comment (mattcrn) [reply includes/FailedRequestQueue/RetryFailedRequestsQueueWorker.php:99]: or is it just in case?

File: tests/Unit/CreateLogisticsOrderCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        $this->order->shouldHaveReceived("save");
+    }
+
+    public function test_it_does_not_update_the_order_state_after_a_failed_request()
+    {
+        /** @var MockInterface|WriteApi */
+        $writeApi = Mockery::mock(WriteApi::class);
+        $writeApi->shouldReceive("logisticsOrderPost")->andThrow(new ApiException("Unauthenticated", 401));
+
+        try {
```
Comment (mattcrn) [inline tests/Unit/CreateLogisticsOrderCommandTest.php:68]: You could use phpunits `expectException`

Comment (mattcrn) [reply tests/Unit/CreateLogisticsOrderCommandTest.php:68]: oh, you used it below, nvm

**Review by mattcrn — APPROVED** [review]

File: includes/FailedRequestQueue/RetryFailedRequestsQueueWorker.php (hunk 4 above)
Comment (lucapircher) [reply includes/FailedRequestQueue/RetryFailedRequestsQueueWorker.php:99]: It could be null if the order was deleted.
